[4주차] 조성아 과제 제출합니다.

sungahChooo opened the pull request:
## [🔗배포주소](https://react-messenger-22nd-git-sungahchooo-chosungahs-projects.vercel.app/)
## ✨결과 화면
<img width="1895" height="860" alt="스크린샷 2025-10-25 233311" src="https://github.com/user-attachments/assets/51c02041-c75a-42fa-b641-75bd88a62d44" />
친구 목록 화면

<img width="1898" height="868" alt="스크린샷 2025-10-25 233301" src="https://github.com/user-attachments/assets/c4630ec6-e437-429f-9fc2-bc5bbf18f169" />
채팅방 목록 화면

<img width="1918" height="865" alt="스크린샷 2025-10-25 233351" src="https://github.com/user-attachments/assets/dc05ed3d-1412-4544-9497-05ff315b38dd" />
프로필 화면

<img width="1918" height="857" alt="스크린샷 2025-10-25 233403" src="https://github.com/user-attachments/assets/9e1b625e-65df-4132-a976-10a8c35aa0e0" />
채팅방 화면

<img width="1919" height="863" alt="스크린샷 2025-10-25 233325" src="https://github.com/user-attachments/assets/cfe22dd4-2b6a-4a54-803c-ee9e1239951a" />
개발 안할 페이지

## 🎨디자이너 피드백 반영 사항
- 채팅버블 정렬
- 시간 font-weight 수정 
- 채팅방 헤더 아이콘 gap조절
- 버블 최대 너비 조절

## ✅지난 과제 이후 구현 기능
- 채팅 버블 공감하기 버튼(hover시 공감 버튼이 나타납니다.)
- 채팅방 상단 고정 기능(채팅방 목록 옆 pin버튼 있습니다.)
- 친구 프로필 사진이  보
- 친구 프로필 조회 기능
- 통화, 커뮤니티,설정 부분 페이지 페이지 개발
- 프로필 배경 이미지 크기 조절
- 상단바 넣기(시간만 현재시간을 불러왔고 나머지 아이콘들은 하드 코딩을 하였습니다.)
- 시간 중복되면 한번만 표시하기
- 오늘 날짜 불러와서 표시
- 채팅 목록에서 안 읽은 메시지 스타일링 적용
- 헤더부분 투명도 조절

## 👍개발 측면에서 개선 사항
- 파일 나누기 리팩토링하기
- 업데이트 주기 설정
- 절대 경로 적용
- svg 색깔 변경하기  
- zustand를 적용

## Review Questions
### 1. React Router의 동적 라우팅(Dynamic Routing)이란 무엇이며, 언제 사용하나요?
정적으로 미리 정의된 경로만 사용하는 것이 아니라, 사용자의 행동이나 서버 데이터에 따라 라우트를 결정하는 것이다.
예를 들어, 내가 프로필을 클릭하면 프로필 페이지로 이동하는 것을 말한다.

### 2. 네트워크 속도가 느린 환경에서 사용자 경험을 개선하기 위해 사용할 수 있는 UI/UX 디자인 전략과 기술적 최적화 방법은 무엇인가요?
UI/UX 디자인 전략:
로딩 상태 표시: 스켈레톤 화면, 스피너, 로딩 바 등으로 사용자에게 진행 상황 제공
점진적 로딩(Progressive Loading): 중요한 콘텐츠부터 먼저 보여주기

기술적 최적화 방법:
코드 스플리팅(Code Splitting): React의 React.lazy, Suspense 사용하여 필요한 컴포넌트만 로딩
이미지 최적화: WebP, AVIF 포맷, Lazy Loading, CDN 활용
네트워크 요청 최소화: API 합치기, GraphQL 사용, 캐싱 전략 적용

### 3. React에서 useState와 useReducer를 활용한 지역 상태 관리와 Context API 및 전역 상태 관리 라이브러리의 차이점을 설명하세요.
#### 지역 상태 관리: useState와 useReducer
useState
용도: 단순하고 독립적인 컴포넌트 상태를 관리할 때 사용
특징: 간단한 boolean, string, number, 객체 상태 관리
상태 업데이트 시 해당 컴포넌트만 리렌더링

useReducer
용도: 상태 전환 로직이 복잡하거나 여러 액션이 있을 때
특징: 상태와 상태 변경 로직을 분리 → 가독성 ↑
여러 액션을 한 곳에서 관리 가능

Context API
용도: 컴포넌트 트리 전체에서 전역적으로 상태 공유가 필요할 때
특징: createContext + Provider를 사용해 트리 전체에 상태 제공
useContext로 어디서든 접근 가능

#### 전역 상태 관리 라이브러리 (예: Redux, Zustand, Recoil)
용도: 대규모 앱에서 복잡한 전역 상태 관리
특징: 상태, 액션, 사이드 이펙트 관리가 명확
특정 컴포넌트만 선택적으로 구독 → 불필요한 리렌더 방지
미들웨어 지원 → 비동기 처리, 로깅, 캐싱 용이

sungahChooo committed:
chore: 개발 환경 설정

sungahChooo committed:
chore: 불필요한 파일 삭제

sungahChooo committed:
chore: 불필요한 파일 삭제

sungahChooo committed:
feat: friendList페이지 구현, 사이드바 구현

sungahChooo committed:
fix: sidebar 버그 수정

sungahChooo committed:
feat: 채팅방 조회 페이지 스타일링

sungahChooo committed:
feat: 채팅방 스타일링 구현

sungahChooo committed:
fix: sidebar 스타일링 수정

sungahChooo committed:
feat: profile page 구현

sungahChooo committed:
fix: 친구 프로필 스타일 수정, 채팅방 목록 스타일 보완

sungahChooo committed:
feat:채팅 기능 구현

sungahChooo committed:
fix: 배포 오류 수정

sungahChooo committed:
fix: App.tsx 오타 수정

sungahChooo committed:
fix: 채팅방 친구 프로필과 이름 추가

sungahChooo committed:
fix: 배포 오류 해결

sungahChooo committed:
fix: 디자이너 피드백 반영

sungahChooo committed:
fix: 한글 끝글자 중복 입력 버그 해결

sungahChooo committed:
fix: 한글 끝글자 중복 입력 방지 처리

sungahChooo committed:
fix: 한글 끝글자 중복 입력 방지

sungahChooo committed:
fix: 한글 중복 방지 처리

sungahChooo committed:
feat: 공감기능 구현

sungahChooo committed:
design: favicon적용

sungahChooo committed:
fix: 아이콘 깨짐 해결 및 svg파일 색 변경 가능하도록 수정

sungahChooo committed:
fix: QA 피드백 반영 – 버튼 간격 및 정렬 수정

sungahChooo committed:
chore: 절대경로 적용

sungahChooo committed:
feat: 개발 안할 페이지 스타일링 및 개발

sungahChooo committed:
feat: 상단바 기능 구현

sungahChooo committed:
feat: 친구 프로필 조회 기능 구현

sungahChooo committed:
fix: 친구 목록 조회 오류 수정

sungahChooo committed:
fix: 배포를 위한 오류 수정

sungahChooo committed:
feat: 채팅방 상단 고정, 안읽은 메시지 표시, 채팅 상대 이름 표시

sungahChooo committed:
fix: 채팅방 목록 페이지 배열 조절

sungahChooo committed:
fix: status bar 투명도 조절

sungahChooo committed:
fix: header부분 투명도 조절

sungahChooo committed:
refact: 컴포넌트화 리팩토링

sungahChooo committed:
fix: 채팅 페이지 오류 해결

sungahChooo committed:
fix: 채팅방 스크롤 문제 해결

sungahChooo committed:
fix: 채팅방 시간 띄우기

sungahChooo committed:
fix: layout 오류 해결

sungahChooo committed:
refact: 프로필 페이지 컴포넌트화

sungahChooo committed:
refact: 친구 프로필 컴포넌트화

sungahChooo committed:
feat: 채팅방 목록 친구 이름 띄우기 및 채팅 더미데이터 수정

sungahChooo committed:
fix: 채팅방 목록 프로필 이미지 오류 수정

sungahChooo committed:
refact: 채팅방 목록 프로필 이미지 컴포넌트화

sungahChooo committed:
refact: 채팅방 리팩토링

sungahChooo committed:
fixt: 채팅방 이미지 오류 해결

sungahChooo committed:
fix: 채팅방 시간 형식 수정

sungahChooo committed:
feat: 채팅방 날짜 화면 표시 기능

sungahChooo committed:
fix: 채팅방 날짜 표시 오류 해결

sungahChooo committed:
feat: 채팅방 메시지 최근 시간만 표시

sungahChooo committed:
로컬 스토리지 초기화 코드 삭제

sungahChooo committed:
util함수 오류 해결

only1Ksy commented on src/App.tsx:
```diff
@@ -0,0 +1,48 @@
+import Statusbar from '@/components/Statusbar';
+import FriendProfile from '@/pages/friendList/profile/FriendProfile';
+
+function AppLayout({ children }: { children: React.ReactNode }) {
```
layout은 따로 컴포넌트로 분리해 사용하면 좋을 것 같아요!

only1Ksy commented on src/components/Navbar.tsx:
```diff
@@ -0,0 +1,88 @@
+  return (
+    <nav className="border-light-gray space-between fixed bottom-0 left-1/2 z-900 flex h-[65px] w-full max-w-[375px] -translate-x-1/2 items-center justify-around border-t bg-white">
+      <div className="flex w-[46px] flex-col items-center justify-center border-0 bg-transparent">
+        <Link to="/friendList">
```
현재 아이콘/텍스트 구성이 반복되는 형태로 보이는데, 이는 배열 + map 렌더링으로 리팩토링하시면 유지보수에 더욱 효과적일 것 같습니당!

only1Ksy commented on src/components/Navbar.tsx:
```diff
@@ -0,0 +1,88 @@
+
+import { Link, useLocation } from 'react-router-dom';
+
+function Sidebar() {
```
파일명은 Navbar이고 함수명은 Sidebar인데, 특별한 이유가 없다면 통일하는 것이 좋을 것 같습니다!

sungahChooo replied:
네 반영했어요~

only1Ksy commented on src/pages/calls/Call.tsx:
지금 pages 폴더 안에 components와 pages가 함께 있는 것 같은데, pages와 components로 폴더 분리하시면 더 좋을 것 같습니당

sungahChooo replied:
넵 반영했습니다~

only1Ksy commented on src/data/chat.json:
```diff
@@ -0,0 +1,146 @@
+        "roomId": "1",
+        "sender": 2,
+        "message": "안녕하세요! Room 1입니다.",
+        "time": "3:00 PM",
```
지금 mock data의 time 필드가 위 형식으로 채워져 있어서 화면에서 Invalid Date로 뜨고 있습니다!

only1Ksy commented on src/components/chat/ChatListItem.tsx:
```diff
@@ -0,0 +1,45 @@
+              />
+            </button>
+          </p>
+          <span className="text-xs font-normal text-gray-600">{room.messages[0].time}</span>
```
여기서 표시되고 있는 시간이 메시지 배열의 첫 번째 메시지 시간으로, 잘못되어 있는 것 같습니다!
그리고, 처음 mock data를 설정할 때 last message로 설정했던 게 채팅을 보내는 과정에서 바뀌지 않아 화면에서도 반영되지 않는 것 같습니다!
-> 그래서 결국 pin 정렬 로직도 반영이 제대로 되지 않아서, 핀을 했다가 해제하면 그냥 제일 최근에 핀 했던 채팅방이 가장 위로 올라와 있는 것을 확인할 수 있습니다 (초기 message 가 invalidate 형식이라 정렬 로직 작동 안 하는 것으로 보임)

only1Ksy commented on src/components/chat/chatRoom/ChatMsgBubble.tsx:
```diff
@@ -0,0 +1,69 @@
+              className={`flex cursor-pointer items-center gap-1 rounded-full bg-gray-200 px-2 py-1 text-xs ${
+                message.likes === 0 ? 'opacity-0 group-hover:opacity-100' : 'opacity-100'
+              }`}
+              onClick={onLike}
```
이 기능까지 구현하셨네요! 👍

only1Ksy commented on src/components/chat/ChattingHeader.tsx:
```diff
@@ -0,0 +1,26 @@
+      right={
+        <div className="flex cursor-pointer gap-4">
+          <button onClick={() => alert('준비중입니다.')}>
+            <img src={searchIcon} alt="search" />
```
위에서 활용하셨던 것처럼, Icon svg 파일들을 컴포넌트화 해서 사용하시면 좋을 것 같습니다!

only1Ksy commented on src/pages/chat/chattingRoom/ChattingRoom.tsx:
```diff
@@ -0,0 +1,121 @@
+    prevLengthRef.current = messages.length;
+  }, [messages]);
+
+  const getFormattedDate = (date = new Date()) =>
```
이런 로직은 유틸 함수로 따로 분리하는 것도 좋을 것 같습니다!

only1Ksy commented on src/pages/chat/chattingRoom/ChattingRoom.tsx:
```diff
@@ -0,0 +1,121 @@
+            const sender = users.find((u) => u.id === msg.sender);
+            return (
+              <ChatMessageBubble
+                key={idx}
```
key 값으로 index를 사용하는 것은 권장되지 않는 방식입니다! id 같은 고유한 값으로 설정해 주세용

[리액트에서 key에 index를 넣으면 안 되는 이유](https://yozm.wishket.com/magazine/detail/2634/)

sungahChooo replied:
반영했습니다~!

only1Ksy commented on src/components/chat/chatRoom/ChattingRoomHeader.tsx:
```diff
@@ -0,0 +1,49 @@
+            <img src={backIcon} alt="back" />
+          </span>
+          <ChattingListImgSection participants={participants} />
+          <span>{names}</span>
```
<img width="400" height="911" alt="스크린샷 2025-10-26 오전 12 51 02" src="https://github.com/user-attachments/assets/6cccfe8d-be89-4caf-904c-f21ca34b0fea" />
대화 상대가 많거나 이름이 길어지면 위와 같이 줄바꿈 되네요! 이 부분에 대한 처리가 필요할 것 같습니당

sungahChooo replied:
해결했습니다!

only1Ksy commented on src/pages/friendList/profile/FriendProfile.tsx:
이 컴포넌트와 MyProfile 컴포넌트는 같은 틀 안에서 사용되는 데이터만 다른 것 같아요! 데이터를 주입해 사용할 수 있는 구조로 리팩토링하면 중복되는 코드를 줄일 수 있을 것 같습니당

only1Ksy reviewed (commented):
안녕하세요, 성아 님! 이번 주차 과제도 수고 많으셨습니당 👍👏

sungahChooo committed:
refact: 컴포넌트 폴더 주소 변경

Wannys26 commented on src/stores/chatStore.ts:
```diff
@@ -0,0 +1,69 @@
+          return { chatRooms: sortChatRooms(updated) };
+        }),
+
+      setCurrentRoom: (roomId) => {
```
messages라는 단일 배열 안에 메시지, 좋아요 등등이 모두 저장되어있고, setCurrentRoom을 통해서 채팅방에서 보낸 메시지, 좋아요 등을 저장하는거 같은데,
지금 chatRoom1에서 채팅을 하다가 2로 넘어가는 상황이 발생하면, setCurrentRoom이라는 함수를 통해서 chatRoom1에서 발생한 메시지, 좋아요 내역들이 전부 삭제되는 거 같습니다! 
addMessage나 toggleLike 함수에서 messages 단일 배열에 모두 저장하기 보다는, chatRoom id 별로 구분해서 저장하고 채팅방을 이동하면 불러오는 식으로 수정하면 좋을 거 같습니다!

Wannys26 commented on src/components/friendList/MyProfileSection.tsx:
```diff
@@ -0,0 +1,21 @@
+    >
+      <img src={profile} alt="프로필 이미지" className="h-[65px] w-[65px]" />
+      <div>
+        <div className="text-2xl font-semibold">정해인</div>
```
여기 하드 코딩 하기보다는, 작성하신 JSON 파일에서 불러오는 방식으로 수정하면 좋을거 같습니다!

Wannys26 commented on src/components/friendList/FriendListSection.tsx:
```diff
@@ -0,0 +1,53 @@
+
+export default function FriendListSection() {
+  // 이미지 매핑 객체
+  const images: Record<string, string> = {
```
data/ProfileImages.ts에 이미 images 매핑하는 함수가 있으니, 이거 그대로 사용하시는게 좋을거 같습니다!

Wannys26 commented on src/pages/friendList/profile/FriendProfile.tsx:
```diff
@@ -0,0 +1,54 @@
+import ProfileCard from './ProfileBox';
+
+// 이미지 매핑
+const images: Record<string, string> = {
```
data/ProfileImages.ts에 이미 images 매핑하는 함수가 있으니, 이거 그대로 사용하시는게 좋을거 같습니다!

Wannys26 commented on src/stores/chatStore.ts:
```diff
@@ -0,0 +1,69 @@
+        });
+      },
+
+      addMessage: (message) => {
```
채팅방에서 채팅을 보내도 채팅 목록 페이지에서 최신 메시지가 업데이트 되지 않는거 같습니다! 메시지를 보낼 때마다, 최근에 보낸 메시지가 업데이트 되도록 수정해야 될거 같습니다

Wannys26 reviewed (commented):
코드 잘 봤습니다! 이번 과제도 수고하셨고, 시험 치느라 고생많으셨습니다!
다음주도 좋은 코드 기대하겠습니다!

baekseungsun commented on src/components/warning.tsx:
오류 Warning을 따로 준비하신 게 인상적입니다!

baekseungsun commented on src/components/chat/ChatListItem.tsx:
직접 전송한 메세지 말고 Mock 메세지는 시간이 invalid date으로 뜨는 문제가 있는 것 같습니다

sungahChooo replied:
해결했습니다 ㅎㅎ

baekseungsun commented on src/pages/chat/chattingRoom/ChattingRoom.tsx:
```diff
@@ -0,0 +1,121 @@
+
+        {/* 오늘 메시지 */}
+        {todayMessages.map((msg, idx) => {
+          const sender = users.find((u) => u.id === msg.sender);
```
const sender = users.find((u)는 매 유저를 순회해야 하기에 유저가 많이 늘어나면 시간이 오래 걸릴 것 같습니다. 매핑을 사용하는 것도 괜찮아 보입니다

baekseungsun commented on src/pages/friendList/profile/MyProfile.tsx:
```diff
@@ -0,0 +1,30 @@
+
+const myButtons = [
+  { icon: profileChatIcon, label: '나와의 채팅', onClick: () => alert('채팅 준비중') },
+  { icon: editIcon, label: '프로필 편집', onClick: () => alert('편집 준비중') },
```
Alert로 채팅 준비중, 편집 준비중 보여주시는 게 유저 입장에서 좋은 것 같습니다

sungahChooo committed:
fix: 채팅방 더미데이터에서 불러온 메시지 시간 안뜸  오류 해결

sungahChooo committed:
fix: 오류 방지를 위한 채팅버블 key값 수정

sungahChooo committed:
style: 채팅방 헤더 참여자 이름 최대 크기 지정

sungahChooo committed:
refactor: 전체 레이아웃 수정

sungahChooo committed:
Add initial project documentation to README

sungahChooo committed:
Add layout section to README

Added layout section explaining the structure and optimization of the application.

sungahChooo committed:
refact: layout 재구성

sungahChooo committed:
Merge branch 'sungahChooo' of https://github.com/sungahChooo/react-messenger-22nd into sungahChooo
